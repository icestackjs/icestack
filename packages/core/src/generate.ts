import * as t from '@babel/types'
import _babelGenerate from '@babel/generator'
import { stages } from '@icestack/shared/constants'
// https://github.com/babel/babel/issues/15269
function _interopDefaultCompat(e: any) {
  return e && typeof e === 'object' && 'default' in e ? e.default : e
}
export const babelGenerate = _interopDefaultCompat(_babelGenerate) as typeof _babelGenerate

export function generateIndexCode(basenames: string[], type: 'components' | 'utilities' = 'components') {
  if (type === 'components') {
    const props = Object.entries(
      basenames.reduce<Record<string, t.ObjectProperty[]>>((acc, name) => {
        for (const stage of stages) {
          const node = t.objectProperty(t.stringLiteral(stage), t.callExpression(t.identifier('require'), [t.stringLiteral(`./${name}/${stage}.cjs`)]))
          if (acc[name]) {
            acc[name].push(node)
          } else {
            acc[name] = [node]
          }
        }
        return acc
      }, {})
    ).map(([key, props]) => {
      return t.objectProperty(t.stringLiteral(key), t.objectExpression(props))
    })

    const ast = t.file(
      t.program([t.expressionStatement(t.assignmentExpression('=', t.memberExpression(t.identifier('module'), t.identifier('exports')), t.objectExpression(props)))])
    )

    return babelGenerate(ast).code
  } else {
    const props = basenames.map((basename) => {
      return t.objectProperty(t.stringLiteral(basename), t.callExpression(t.identifier('require'), [t.stringLiteral(`./${basename}.cjs`)]))
    })

    const ast = t.file(
      t.program([t.expressionStatement(t.assignmentExpression('=', t.memberExpression(t.identifier('module'), t.identifier('exports')), t.objectExpression(props)))])
    )

    return babelGenerate(ast).code
  }
}

export function generateCva({ format }: { format: 'js' | 'ts' }) {}
