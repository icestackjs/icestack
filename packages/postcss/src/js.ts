// import postcssJs from 'postcss-js'
// import type { AcceptedPlugin } from 'postcss'
// import { TailwindcssPluginOptions } from '@/types'
// options: TailwindcssPluginOptions

const noop: <T>(x: T) => T = (x) => x
export function getJsProcess() {
  // const { runtime } = options
  // const basePlugins: AcceptedPlugin[] = []
  // const componentsPlugins: AcceptedPlugin[] = []
  // const utilitiesPlugins: AcceptedPlugin[] = []

  return {
    baseProcess: noop, // postcssJs.sync(basePlugins),
    componentsProcess: noop, // postcssJs.sync(componentsPlugins),
    utilitiesProcess: noop //  postcssJs.sync(utilitiesPlugins)
  }
}
