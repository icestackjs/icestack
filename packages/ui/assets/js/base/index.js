module.exports = {
  ":root": {
    "--primary-color": "255 255 255",
    "--primary": "25 137 250",
    "--success-color": "255 255 255",
    "--success": "7 193 96",
    "--error-color": "255 255 255",
    "--error": "238 10 36",
    "--warning-color": "255 255 255",
    "--warning": "255 151 106"
  }
}