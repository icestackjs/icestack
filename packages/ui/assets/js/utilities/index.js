module.exports = {
  ".mp-reset-button": {
    "padding": "0",
    "backgroundColor": "transparent",
    "fontSize": "inherit",
    "lineHeight": "inherit",
    "color": "inherit",
    "borderWidth": "0"
  },
  ".mp-reset-button::after": {
    "border": "none"
  }
}