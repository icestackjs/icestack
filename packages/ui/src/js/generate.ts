import path from 'node:path'
import * as t from '@babel/types'
import _babelGenerate from '@babel/generator'
// https://github.com/babel/babel/issues/15269
function _interopDefaultCompat(e: any) {
  return e && typeof e === 'object' && 'default' in e ? e.default : e
}
export const babelGenerate = _interopDefaultCompat(_babelGenerate) as typeof _babelGenerate

export function generateIndexCode(basenames: string[]) {
  // const booleanMap: Record<string, boolean> = {}
  const props = Object.entries(
    basenames
      .map((basename) => {
        const s = basename.split(path.sep)
        return {
          key: s[0],
          type: s[1],
          path: basename
        }
      })
      .reduce<Record<string, t.ObjectProperty[]>>((acc, { key, path, type }) => {
        const node = t.objectProperty(t.stringLiteral(type), t.callExpression(t.identifier('require'), [t.stringLiteral(`./${path}.js`)]))
        if (acc[key]) {
          acc[key].push(node)
        } else {
          acc[key] = [node]
        }
        return acc
      }, {})
  ).map(([key, props]) => {
    return t.objectProperty(t.stringLiteral(key), t.objectExpression(props))
  })

  const ast = t.file(
    t.program([t.expressionStatement(t.assignmentExpression('=', t.memberExpression(t.identifier('module'), t.identifier('exports')), t.objectExpression(props)))])
  )

  return babelGenerate(ast).code
}

export function generateComponentsIndexCode(basenames: string[]) {
  const props = Object.entries(
    basenames.reduce<Record<string, t.ObjectProperty[]>>((acc, name) => {
      for (const stage of ['base', 'styled', 'utils']) {
        const node = t.objectProperty(t.stringLiteral(stage), t.callExpression(t.identifier('require'), [t.stringLiteral(`./${name}/${stage}.js`)]))
        if (acc[name]) {
          acc[name].push(node)
        } else {
          acc[name] = [node]
        }
      }
      return acc
    }, {})
  ).map(([key, props]) => {
    return t.objectProperty(t.stringLiteral(key), t.objectExpression(props))
  })

  const ast = t.file(
    t.program([t.expressionStatement(t.assignmentExpression('=', t.memberExpression(t.identifier('module'), t.identifier('exports')), t.objectExpression(props)))])
  )

  return babelGenerate(ast).code
}
