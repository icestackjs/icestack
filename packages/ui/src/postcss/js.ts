import postcssJs from 'postcss-js'
import type { AcceptedPlugin } from 'postcss'
// import { getPlugin } from './custom-property-prefixer'
import postcssPrefix from './prefixer'
import globalPostcss from './global'
// import { rtlcss } from './rtlcss'
// import basePlugin from './runtime/base'
import { CodegenOptions } from '@/types'

export function getJsProcess(options: CodegenOptions) {
  const { runtime } = options
  const basePlugins: AcceptedPlugin[] = []
  const componentsPlugins: AcceptedPlugin[] = []
  const utilitiesPlugins: AcceptedPlugin[] = []

  // if (varPrefix) {
  //   const p = getPlugin(varPrefix)
  //   basePlugins.push(p)
  //   componentsPlugins.push(p)
  //   utilitiesPlugins.push(p)
  // }
  const { prefix } = runtime
  if (prefix) {
    const p = postcssPrefix(typeof prefix === 'string' ? { prefix, ignore: [] } : prefix)
    componentsPlugins.push(p)
    utilitiesPlugins.push(p)
  }
  // if (rtl) {
  //   const p = rtlcss(typeof rtl === 'boolean' ? undefined : rtl)
  //   componentsPlugins.push(p)
  //   utilitiesPlugins.push(p)
  // }

  if (runtime) {
    const p = globalPostcss(options)
    basePlugins.push(p)
    componentsPlugins.push(p)
    utilitiesPlugins.push(p)
  }

  // if (base) {
  //   const p = basePlugin(options)
  //   basePlugins.push(p)
  // }

  return {
    baseProcess: postcssJs.sync(basePlugins),
    componentsProcess: postcssJs.sync(componentsPlugins),
    utilitiesProcess: postcssJs.sync(utilitiesPlugins)
  }
}
