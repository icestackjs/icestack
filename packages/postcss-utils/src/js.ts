import type { CssInJs } from 'postcss-js'
import postcssJs from 'postcss-js'
import type { AcceptedPlugin, Document, LazyResult, Root } from 'postcss'
import postcss from 'postcss'

const noop: <T>(x: T) => T = x => x
export function getJsProcess() {
  return {
    baseProcess: noop, // postcssJs.sync(basePlugins),
    componentsProcess: noop, // postcssJs.sync(componentsPlugins),
    utilitiesProcess: noop, //  postcssJs.sync(utilitiesPlugins)
  }
}

export function objectify(root: Root) {
  return postcssJs.objectify(root as Root)
}

export function parseJs(cssInJs: CssInJs) {
  return postcssJs.parse(cssInJs)
}

export function stringify(cssInJs: CssInJs) {
  return parseJs(cssInJs).toString()
}

export function postcssProcess(plugins: AcceptedPlugin[], css: string): LazyResult<Document | Root> {
  // @ts-ignore
  return postcss(plugins).process(css, {
    from: undefined,
  })
}

export { generateCva } from 'postcss-cva/generator'
