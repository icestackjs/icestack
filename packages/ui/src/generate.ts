import fs from 'node:fs/promises'
import path from 'node:path'
import plugin from 'tailwindcss/plugin'
import { set } from 'lodash'
import klaw from 'klaw'
import defu from 'defu'
import { generateIndexCode } from './js/generate'
import { buildScss } from '@/sass'
import { resolveJsDir, scssDir } from '@/dirs'
import { defaultVarPrefix, someExtends } from '@/constants'
export interface IOptions {
  dir?: string
  outSideLayerCss: 'base' | 'utilities' | 'components'
}

export async function generate(options: IOptions) {
  const { outSideLayerCss, dir } = defu<IOptions, Partial<IOptions>[]>(options, {})
  // await ensureDir(pluginsDir)
  switch (outSideLayerCss) {
    case 'base': {
      for await (const file of klaw(path.resolve(scssDir, 'base'))) {
        await buildScss({
          dir,
          filename: file.path,
          stats: file.stats,
          outSideLayerCss
        })
      }

      break
    }
    case 'utilities': {
      const { colors } = await import('../src/colors')
      const utilitiesPath = path.resolve(scssDir, 'utilities')
      const basenameArray = []
      const fromDir = path.resolve(scssDir, 'utilities')
      const utilitiesJsOutputPath = path.resolve(resolveJsDir(dir), 'utilities')
      for await (const file of klaw(path.resolve(utilitiesPath, 'global'))) {
        if (file.stats.isFile() && /\.scss$/.test(file.path)) {
          if (path.basename(file.path).startsWith('_')) {
            continue
          }
          basenameArray.push(path.relative(fromDir, file.path).replace(/\.scss$/, ''))
          await buildScss({
            dir,
            filename: file.path,
            stats: file.stats,
            outSideLayerCss,
            resolveConfig(config) {
              set(config, 'theme.extend.colors', colors)
            }
          })
        }
      }
      await fs.writeFile(
        path.resolve(utilitiesJsOutputPath, 'index.js'),
        generateIndexCode(basenameArray),
        'utf8'
        // dedent`module.exports = {\n${basenameArray.map((x) => {
        //   return `"${x}":require("./${x}.js")`
        // })}\n}`
      )
      basenameArray.length = 0
      const utilitiesJs = path.resolve(resolveJsDir(dir), 'utilities')

      for await (const file of klaw(utilitiesPath)) {
        if (file.stats.isFile() && /\.scss$/.test(file.path)) {
          if (path.basename(file.path).startsWith('_')) {
            continue
          }

          basenameArray.push(path.relative(fromDir, file.path).replace(/\.scss$/, ''))
          await buildScss({
            dir,
            filename: file.path,
            stats: file.stats,
            outSideLayerCss,
            resolveConfig(config) {
              set(config, 'theme.extend.colors', colors)
              config.plugins = [
                plugin(
                  ({ addUtilities }) => {
                    const obj = require(path.resolve(utilitiesJs, 'index.js'))
                    // @ts-ignore
                    addUtilities(Object.values(obj))
                  },
                  {
                    theme: {
                      extend: {
                        ...someExtends
                      }
                    }
                  }
                )
              ]
            }
          })
        }
      }

      await fs.writeFile(
        path.resolve(utilitiesJsOutputPath, 'index.js'),
        generateIndexCode(basenameArray),
        'utf8'
        // dedent`module.exports = {\n${basenameArray.map((x) => {
        //   return `"${x}":require("./${x}.js")`
        // })}\n}`
      )

      break
    }
    case 'components': {
      const { colors } = await import('../src/colors')
      const utilitiesJs = path.resolve(resolveJsDir(dir), 'utilities')

      const basenameArray = []
      const fromDir = path.resolve(scssDir, 'components')
      for await (const file of klaw(fromDir)) {
        if (file.stats.isFile() && /\.scss$/.test(file.path)) {
          if (path.basename(file.path).startsWith('_')) {
            continue
          }
          basenameArray.push(path.relative(fromDir, file.path).replace(/\.scss$/, ''))
          await buildScss({
            dir,
            filename: file.path,
            stats: file.stats,
            resolveConfig: (config) => {
              set(config, 'theme.extend.colors', colors)
              config.plugins = [
                plugin(
                  ({ addUtilities }) => {
                    const obj = require(path.resolve(utilitiesJs, 'index.js'))
                    // @ts-ignore
                    addUtilities(Object.values(obj))
                  },
                  {
                    theme: {
                      extend: {
                        ...someExtends
                      }
                    }
                  }
                )
              ]
            },
            outSideLayerCss
          })
        }
      }
      const componentsJsOutputPath = path.resolve(resolveJsDir(dir), 'components')
      await fs.writeFile(path.resolve(componentsJsOutputPath, 'index.js'), generateIndexCode(basenameArray), 'utf8')
      break
    }
    default:
  }
}

export async function buildAll(dir?: string) {
  await generate({
    dir,
    outSideLayerCss: 'base'
  })
  await generate({
    dir,
    outSideLayerCss: 'utilities'
  })
  await generate({
    dir,
    outSideLayerCss: 'components'
  })
}
