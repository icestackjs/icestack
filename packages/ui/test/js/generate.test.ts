import { generateIndexCode } from '@/js/generate'
describe('generate', () => {
  it('case 0', () => {
    expect(generateIndexCode(['global/glass', 'styled/artboard', 'unstyled/button'])).toMatchSnapshot()
  })

  it('case 1', () => {
    expect(generateIndexCode(['global/glass', 'styled/glass', 'unstyled/glass'])).toMatchSnapshot()
  })
})
