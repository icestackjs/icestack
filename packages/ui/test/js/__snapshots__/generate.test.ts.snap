// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generate > case 0 1`] = `
"module.exports = {
  \\"glass\\": {
    \\"global\\": require(\\"./global/glass.js\\")
  },
  \\"artboard\\": {
    \\"styled\\": require(\\"./styled/artboard.js\\")
  },
  \\"button\\": {
    \\"unstyled\\": require(\\"./unstyled/button.js\\")
  }
};"
`;

exports[`generate > case 1 1`] = `
"module.exports = {
  \\"glass\\": {
    \\"global\\": require(\\"./global/glass.js\\"),
    \\"styled\\": require(\\"./styled/glass.js\\"),
    \\"unstyled\\": require(\\"./unstyled/glass.js\\")
  }
};"
`;
